File: src/main/java/com/github/zxh/classpy/classfile/jvm/Mutf8Decoder.java
package com.github.zxh.classpy.classfile.jvm;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

public class Mutf8Decoder {
    
    /**
     * Decode modified UTF-8 string from byte[].
     * todo: optimize
     * 
     * @param bytes
     * @return 
     * @throws java.io.IOException 
     */
    public static String decodeMutf8(byte[] bytes) throws IOException {
        ByteArrayOutputStream baos = new ByteArrayOutputStream(bytes.length + 2);
        DataOutputStream dos = new DataOutputStream(baos);
        dos.writeShort(bytes.length);
        dos.write(bytes);

        ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
        DataInputStream dis = new DataInputStream(bais);
        return dis.readUTF();
    }
    
}

File: src/main/kotlin/com/github/zxh/classpy/helper/Mutf8Decoder.kt
package com.github.zxh.classpy.helper

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

object Mutf8Decoder {
    
    /**
     * Decode modified UTF-8 string from byte[].
     * todo: optimize
     * 
     * @param bytes
     * @return 
     * @throws java.io.IOException 
     */
    @Throws(IOException::class)
    @JvmStatic
    fun decodeMutf8(bytes: ByteArray): String {
        val baos = ByteArrayOutputStream(bytes.size + 2);
        val dos = DataOutputStream(baos)
        dos.writeShort(bytes.size);
        dos.write(bytes);

        val bais = ByteArrayInputStream(baos.toByteArray());
        val dis = DataInputStream(bais);
        return dis.readUTF();
    }
    
}
